File: vscodeExt/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
# Time Limit Exceeded
# "aaaaaaaaaaaaaaaaaaa"
class Solution_fail:
    
    def longestPalindrome(self, s: str) -> str:
        if len(s)< 2:
            return s 
        maxL = 1 # 最长子串长度
        startI = 0 # 开始index
        i = 0
        while i < len(s):
            l = 0
            while i + l <= len(s):
                if self.isPalindrome(s[i:i+l]):
                    if l > maxL: 
                        maxL = l 
                        startI = i 
                l+=1
            i+=1
        return s[startI: startI + maxL]

    def isPalindrome(self,s: str):
        if s == s[::-1]:
            return True
        else:
            return False  

class Solution:
    # 超时
    def longestPalindrome_1(self, s: str) -> str:
        if len(s)< 2:
            return s 
        maxL = 1 # 最长子串长度
        startI = 0 # 开始index
        i = 0
        while i < len(s):
            l = 0
            while i + l <= len(s):
                if self.isPalindrome2(s[i:i+l]):
                    if l > maxL: 
                        maxL = l 
                        startI = i 
                l+=1
            i+=1
        return s[startI: startI + maxL]

    # 功能函数在此
    def longestPalindrome(self, s: str) -> str:
        if len(s)< 2:
            return s 
        #记录最终要返回的状态    
        start = 0 # 
        end = 0 # 
        # j = 0
        # 
        # 从0 到 len(s)//2 - 1
        i = 0 # 游标 
        while i < len(s):
            # two cases:
            # 1, 偶数子串            
            j = 0
            while j <= i and i+j+1 < len(s):
                if s[i-j] == s[i+j+1]:
                    if 2*j+2 > end - start:
                        start = i -j 
                        end = i+j+1
                    j += 1
                else:
                    break         
            # 2,奇数子串
            j = 1 
            while j <= i and i+j <len(s):    
                # 共用i,j时 已经变动
                if s[i-j] == s[i+j]:
                    if 2*j + 1 > end - start:
                        start = i-j
                        end = i+j         
                    j += 1
                else:
                    break
            i += 1

        return s[start: end+1]    

    def isPalindrome(self,s: str):
        if s == s[::-1]:
            return True
        else:
            return False  
    # len()为偶数时，比较的长度为 len()//2 - 1 # 4, 比较0，1
    # len()为奇数时，比较的长度为 len()//2 - 1 # 3，比较0
    def isPalindrome2(self,s: str):
        if len(s) < 2:
            return True 
        # 实际取值为0...len(s)//2 - 1    
        for i in range(len(s)//2 ):
            if(s[i] != s[len(s)-1-i]):
                return False
        return True         

# @lc code=end

File: vscodeExt/5.longest-palindromic-substring.java
import java.util.Arrays;

import sun.net.www.content.audio.basic;

/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    // 使用复制传值， 与 实例变量；

    int start = 0;
    int len = 1;
    public String longestPalindrome(String s) {
        // 当为“” 时 subString出错。 
        if( s.length() < 2){
            return s;
        }

        for(int i = 0;i<s.length()-1;i++){
            extend(s,i,i);
            extend(s,i,i+1);
        }
        return s.substring(start, start+len );

    }
    // 向两边扩展
    void extend(String s, int i , int j){
        while(i >=0  && j < s.length() && s.charAt(i) == s.charAt(j)){
            i -= 1;
            j += 1;
        }
        // 此时已经不等
        if( j -i- 1> len){
            len = j -i- 1;
            start = i+1;
        }
    }



    // 分开讨论的
    public String longestPalindrome_1(String s) {
        if(s.length() <2){
            return s;
        }
        char[] sa = s.toCharArray();
        int start= 0;
        int len = 1;
        for(int i=0;i<sa.length;i++){
            //1, 偶数子串
            for(int j =0 ; i-j>=0 && i+j+1<sa.length;j++){
                if(sa[i-j] != sa[i+j+1]){
                    break;
                } else {
                    if (2*j+2 >len){
                        start = i-j;
                        len = 2*j+2;
                    }
                }
            }
            // 2 奇数子串
            for(int j =1; i-j>=0 && i+j <sa.length;j++){
                if(sa[i-j] !=sa[i+j]){
                    break;
                } else {
                    if(2*j +1 >len){
                        start =i-j;
                        len = 2*j +1;
                    }
                }
                // System.out.println(""+start+" "+len);
            }


        }        
        // System.out.println(""+start+" "+len);
        return String.copyValueOf(sa,start,len);
    }




    // Time Limit Exceeded
    public String longestPalindrome_TLE(String s) {
        if(s.length() <2){
            return s;
        }
        char[] sa = s.toCharArray();
        // 需要两层循环, 判断还有一层循环；
        int  maxLen = 1;
        int startI = 0;
        for (int i=0; i < sa.length;i++){
            for(int l = 1; i+l <= sa.length;l++){
                if(isPalindrome(Arrays.copyOfRange(sa, i, i+l))){
                    if (l>maxLen){
                        maxLen = l;
                        startI = i;
                    }
                }
            }
        }
        return String.copyValueOf(sa, startI, maxLen);
        // Arrays.copyOfRange(sa, startI, maxLen));
        
    }

    public boolean isPalindrome(char[] s){
        for(int i =0 ;i< s.length/2;i++){
            if(s[i] != s[s.length -1 -i]){
                return false;
            }
        }
        return true;

    }
}
// @lc code=end

